src/hotspot/share/memory/metadataFactory.hpp

```diff
@@ -65,13 +65,16 @@ class MetadataFactory : AllStatic {
     if (md != NULL) {
       assert(loader_data != NULL, "shouldn't pass null");
       int size = md->size();
-      // Call metadata's deallocate function which will call deallocate fields
+      // Call metadata's deallocate function which will deallocate fields and release_C_heap_structures
       assert(!md->on_stack(), "can't deallocate things on stack");
       assert(!md->is_shared(), "cannot deallocate if in shared spaces");
       md->deallocate_contents(loader_data);
       loader_data->metaspace_non_null()->deallocate((MetaWord*)md, size, md->is_klass());
     }
   }
+
+  // Deallocation method for MethodData because it has a destructor
+  static void free_metadata(ClassLoaderData* loader_data, MethodData* md);
```

Comment:
Rather than providing this overload in order to call the non-trivial ~MethodData(), instead, in the corresponding place in the function template, always call the metadata object's destructor, e.g. md->~T();? If T has an uninteresting destructor then that generates no code, but is okay to call unless the destructor has been deleted (which seems pretty unlikely).

Reply:
I can try that - there was some code in @jcking 's version to get around the C++ that I can borrow.

Reply:
And I took the comment so will make @jcking co-author, when testing is complete.